contracts/BonusRewards.sol

```diff
@@ -3,11 +3,369 @@
+    Pool memory pool = pools[_lpToken];
+    require(pool.lastUpdatedAt > 0, "BonusRewards: pool does not exist");
+    Bonus[] memory bonuses = pool.bonuses;
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      if (bonuses[i].bonusTokenAddr == _bonusTokenAddr) {
+        // when there is alreay a bonus program with the same bonus token, make sure the program has ended properly
+        require(bonuses[i].endTime + WEEK < block.timestamp, "BonusRewards: last bonus period hasn't ended");
+        require(bonuses[i].remBonus == 0, "BonusRewards: last bonus not all claimed");
+      }
+    }
+
+    IERC20 bonusTokenAddr = IERC20(_bonusTokenAddr);
+    uint256 balanceBefore = bonusTokenAddr.balanceOf(address(this));
+    bonusTokenAddr.safeTransferFrom(msg.sender, address(this), _transferAmount);
+    uint256 received = bonusTokenAddr.balanceOf(address(this)) - balanceBefore;
+    // endTime is based on how much tokens transfered v.s. planned weekly rewards
+    uint48 endTime = uint48(received * WEEK / _weeklyRewards + _startTime);
+
+    pools[_lpToken].bonuses.push(Bonus({
+      bonusTokenAddr: _bonusTokenAddr,
+      startTime: _startTime,
+      endTime: endTime,
+      weeklyRewards: _weeklyRewards,
+      accRewardsPerToken: 0,
+      remBonus: received
+    }));
+  }
+
+  /// @notice called by authorizers only, update weeklyRewards (if not ended), or update startTime (only if rewards not started, 0 is ignored)
+  function updateBonus(
+    address _lpToken,
+    address _bonusTokenAddr,
+    uint256 _weeklyRewards,
+    uint48 _startTime
+  ) external override nonReentrant notPaused {
+    require(_isAuthorized(msg.sender, allowedTokenAuthorizers[_lpToken][_bonusTokenAddr]), "BonusRewards: not authorized caller");
+    require(_startTime == 0 || _startTime > block.timestamp, "BonusRewards: startTime in the past");
+
+    // make sure the pool is in the right state (exist with no active bonus at the moment) to add new bonus tokens
+    Pool memory pool = pools[_lpToken];
+    require(pool.lastUpdatedAt > 0, "BonusRewards: pool does not exist");
+    Bonus[] memory bonuses = pool.bonuses;
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      if (bonuses[i].bonusTokenAddr == _bonusTokenAddr && bonuses[i].endTime > block.timestamp) {
+        Bonus storage bonus = pools[_lpToken].bonuses[i];
+        _updatePool(_lpToken); // update pool with old weeklyReward to this block
+        if (bonus.startTime >= block.timestamp) {
+          // only honor new start time, if program has not started
+          if (_startTime >= block.timestamp) {
+            bonus.startTime = _startTime;
+          }
+          bonus.endTime = uint48(bonus.remBonus * WEEK / _weeklyRewards + bonus.startTime);
+        } else {
+          uint256 remBonusToDistribute = (bonus.endTime - block.timestamp) * bonus.weeklyRewards / WEEK;
+          bonus.endTime = uint48(remBonusToDistribute * WEEK / _weeklyRewards + block.timestamp);
```

Comment:
Isn't this problematic if I increase the weekly rewards without changing the balance? Should we add a check to make sure the weeklyRewards isn't increased?

Reply:
why is this problematic? It uses remaining bonus to be distributed to calculate the endtime. If you increase weeklyRewards, you are just shortening the period, less endtime